src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,328 @@ public static CFProxySettings GetSystemProxySettings ()
+			/* CFProxyAutoConfigurationResultCallback __nonnull */ CFProxyAutoConfigurationResultCallbackInternal cb,
+			/* CFStreamClientContext * __nonnull */ ref CFStreamClientContext  clientContext);
+
+		public static CFRunLoopSource ExecuteProxyAutoConfigurationScript (string proxyAutoConfigurationScript, Uri targetUrl, CFProxyAutoConfigurationResultCallback clientCallback, NSObject clientContext)
 		{
 			if (proxyAutoConfigurationScript == null)
-				throw new ArgumentNullException ("proxyAutoConfigurationScript");
+				throw new ArgumentNullException (nameof (proxyAutoConfigurationScript));
 
-			if (targetURL == null)
-				throw new ArgumentNullException ("targetURL");
+			if (targetUrl == null)
+				throw new ArgumentNullException (nameof (targetUrl));
 
-			if (resultCallback == null)
-				throw new ArgumentNullException ("resultCallback");
+			if (clientCallback == null)
+				throw new ArgumentNullException (nameof (clientCallback));
+			
+			using (var pacScript = new NSString (proxyAutoConfigurationScript)) 
+			using (var url = new NSUrl (targetUrl.AbsoluteUri)) {
+				CreatePACCFRunLoopSource factory = delegate (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context) {
+					return CFNetworkExecuteProxyAutoConfigurationScript (pacScript.Handle, url.Handle, cb, ref context);
+				};
+				return ExecutePacCFRunLoopSourceBlocking (factory, clientCallback, clientContext);
+			}
+		}	
+
+		public static CFProxy[] ExecuteProxyAutoConfigurationScript (string proxyAutoConfigurationScript, Uri targetUrl, out NSError outError)
+		{
+			outError = null;
+			IntPtr cbError = IntPtr.Zero;
+			if (proxyAutoConfigurationScript == null)
+				throw new ArgumentNullException (nameof (proxyAutoConfigurationScript));
 
-			if (clientContext == null)
-				throw new ArgumentNullException ("clientContext");
+			if (targetUrl == null)
+				throw new ArgumentNullException (nameof (targetUrl));
 
-			IntPtr source = CFNetworkExecuteProxyAutoConfigurationScript (proxyAutoConfigurationScript.Handle, targetURL.Handle, resultCallback, clientContext);
-			return (source == IntPtr.Zero) ? null : new CFRunLoopSource (source);
+			using (var pacScript = new NSString (proxyAutoConfigurationScript)) 
+			using (var url = new NSUrl (targetUrl.AbsoluteUri)) {
+				CreatePACCFRunLoopSource factory = delegate (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context) {
+					return CFNetworkExecuteProxyAutoConfigurationScript (pacScript.Handle, url.Handle, cb, ref context);
+				};
+				return ExecutePacCFRunLoopSourceBlocking (factory, out outError);
+			}
 		}
 		
+		public static async Task<(CFProxy[] proxies, NSError error)> ExecuteProxyAutoConfigurationScriptAsync (string proxyAutoConfigurationScript, Uri targetUrl, CancellationToken cancellationToken)
```

Comment:
The API takes an `CancellationToken`, but as far as I can see the code doesn't do anything in case of a cancellation request (and the native API isn't cancellable either).

Reply:
The cancellation token is passed to Task.Run, which should cancel the thread unless I am mistaken. It is passed to ExecutePacCFRunLoopSourceAsync and then passed to Task.Run here: https://github.com/xamarin/xamarin-macios/pull/6201/files#diff-edb0955a9801d2d134609eb5bd850840R686

Reply:
No, cancellation is cooperative, nothing will cancel an executing thread (unless you do it yourself). It will however cancel the task if it hasn't been scheduled yet, but that's probably not what users want most of the time.

See also: https://stackoverflow.com/a/22637882/183422